Added Illeism as a speech trait

thetolbean opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
Added a new speech trait, "**illeism**", IE: referring to yourself in third person.

## Why / Balance
This added trait makes roleplaying as an illeist much easier, automatically changing references of I, me, my, and some corresponding phrases, to their correct form, rather than having to do it manually. This saves time and makes roleplaying as an illeist much more appealing to a wider audience. Illeism is compatible with other speech traits, including racial speech patterns. 

## Technical details
This PR adds two new files, `IlleismAccentSystem.cs` & `IlleismAccentComponent.cs`, as well as modifies `traits.ftl` & `speech.yml`.

`IlleismAccentSystem.cs` includes the regex functions to search for and replace the correct phrases.
Phrases include:
I am -> NAME is
I have -> NAME has
I do -> NAME does
I don't -> NAME doesn't
I -> NAME
Me -> NAME
My -> NAME's
_This includes fully capitalized versions_

`IlleismAccentComponent.cs` is the corresponding component.
`traits.ftl` was modified to include a trait name and description.
`speech.yml` was modified to include illeism as a selectable trait worth 1 speech point, as this trait is compatible with others.


## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->


**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->

:cl:
- add: Added Illeism as a selectable speech trait!

Gameandglitch3 commented:
Hell yes.

Errant-4 commented:
Does this unpack the first name, or does it go "Jonathan Verylongname heard you." every time?

lzk228 reviewed (changes requested):
make sure you made all possible strings localizable

Cojoke-dot commented:
maybe I'll -> NAME'll should just break the contraction because I don't think anyone will understand what John'll means. I'll -> NAME will

thetolbean commented:
>1Does this unpack the first name, or does it go "Jonathan Verylongname heard you." every time?

Currently, it does the full name, but I easily could split the name at the first space or dash

thetolbean commented:
The trait now only replaces the phrases with the first part of the name. This split occurs on spaces, not dashes. I wanted to preserve lizard names, ie "eats-the-paper" not becoming "eats", but long names including spaces, ie "Jonathan Verylongname" would become just "Jonathan"

It also does not maintain the contraction I'll, instead replacing it with NAME will

thetolbean commented:
Also, for the traits description, I currently have: _"You seem to only be able to refer to yourself by name."_
If anyone has any ideas to improve it or make it 'quirky' like the other descriptions, I'd love to hear them

Lyroth001 commented:
> Also, for the traits description, I currently have: _"You seem to only be able to refer to yourself by name."_ If anyone has any ideas to improve it or make it 'quirky' like the other descriptions, I'd love to hear them

"You are so full of yourself you feel your actions deserve a narrator, and have taken that task upon yourself. You only refer to yourself in third person" Or something like that?

Nexeron94 commented:
Awesome PR. I would definitely use it on my Vox, which however leads to the question of how very long names and nicknames are handled.

thetolbean commented:
> Awesome PR. I would definitely use it on my Vox, which however leads to the question of how very long names and nicknames are handled.

It currently takes up until the first space, so if you have multiple words in your name, it takes the first.

> "You are so full of yourself you feel your actions deserve a narrator, and have taken that task upon yourself. You only refer to yourself in third person"

I do really like that!

thetolbean commented:
> "You are so full of yourself you feel your actions deserve a narrator, and have taken that task upon yourself. You only refer to yourself in third person" Or something like that?

I like this idea, but it's quite long and extends beyond the boundary of the trait menu, making it seem out of place. Maybe we can shorten it a bit somehow, but keep the general theme?

![image](https://github.com/user-attachments/assets/cf32a7ac-94b2-4767-8326-b9e4fc60a409)

Fildrance commented on Content.Server/Speech/EntitySystems/IlleismAccentSystem.cs:
```diff
@@ -0,0 +1,121 @@
+        message = RegexMyselfUpper.Replace(message, name.ToUpper());
+        if (MostlyUppercase(message))
+        {
+            message = RegexI.Replace(message, name.ToUpper());
```
multiple calls ToUpper not only will result in method call but it also will allocate memory for string every time

thetolbean replied:
Added a var for the uppercase name and just made the strings uppercase rather than relying on ToUpper calls
Code now contains only one ToUpper call

Fildrance commented on Content.Server/Speech/EntitySystems/IlleismAccentSystem.cs:
```diff
@@ -0,0 +1,122 @@
+
+    private void OnAccent(EntityUid uid, IlleismAccentComponent component, AccentGetEvent args)
+    {
+        var entityManager = IoCManager.Resolve<IEntityManager>();
```
how is this thing used?

thetolbean replied:
This entity manager is not being used. It looks like it's a leftover code fragment from a previous implementation i was trying. I'll have to remove it

Fildrance commented on Content.Server/Speech/EntitySystems/IlleismAccentSystem.cs:
```diff
@@ -0,0 +1,122 @@
+    {
+        var entityManager = IoCManager.Resolve<IEntityManager>();
+        var message = args.Message;
+        var name = Name(uid).Split(' ')[0];
```
It'd be nice if we would not create sub strings for each space character in name (use Span magics), but that would be kinda questionable micro-optimization and a lot of additional code for quite rare scenario. I'd better listen to ideas of maintainers on this.

thetolbean replied:
I agree. There's surely a way to only split at the first space rather than every space or to grab up until the first space without splitting at all. The only concern with that, as you brought up, is considering if that tiny optimization makes a big enough difference to warrant the increase in code. I'll brainstorm some options and get maintainer consideration as well.

Fildrance commented on Content.Server/Speech/EntitySystems/IlleismAccentSystem.cs:
```diff
@@ -0,0 +1,122 @@
+    private static readonly Regex RegexMyLower = new(@"\bmy\b|\bmine\b", RegexOptions.IgnoreCase);
+
+    // I'll do it -> NAME'll do it
+    private static readonly Regex RegexIllUpper = new(@"\bI'?LL\b");
```
It might be better to use [source-generated regex](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-source-generators?pivots=dotnet-8-0) but i'm not sure maintainers will be happy about it - it is not used in code base currently.

thetolbean replied:
Yeah, I based my regex strings off of regex strings in other accents. I assume the maintainers would want to keep it standard across the accent system. Maybe a future refactoring of the accent system could improve the regex strings?

thetolbean commented:
Cleaned up leftover code fragment mentioned by @Fildrance and changed the regex string for `I'll`, preventing `ill` from being corrected. IE "He is ill" would previously have been correct to "He is Name will". Now, the apostrophe is required to have it correct. "I don't" still has the apostrophe as optional

thetolbean commented:
Added `I'm` as a corrected phrase to `Name is`.
Now have unrelated test fails :cry:

thetolbean commented:
Unrelated test fails resolved

thetolbean commented:
Did a whole lot of (hopefully) optimizing, just waiting on a review from @Fildrance
Once approved, I can the remove the unused code fragments and commented out code

Fildrance commented:
I am horribly sorry, but i created some benchmark and it seems like Regex.Replace is internally so optimized, that it is better tot be called x2 times then manually doing part of what it is doing to do search once. Even with StringBuilder. Both speed-wise and memory-wise. I honestly am very surprised by that. Benchmark results in following img. ReplaceByHandAppend is your code, ReplaceByRegex is your original code with x2 Regex.Replace calls, and ReplaceByHand is ReplaceByHandAppend but with StringBuilder.Replace instead of Append. 
![image](https://github.com/user-attachments/assets/adbf9a42-c351-4605-ac88-ab8927ef7eae)

It seems like we are reverting some commits :|

Fildrance commented:
Yes, reverted version looks ok, as i remarked earlier - benchmarks show nice time and less allocations...

Fildrance reviewed (approved):
Seems legit!

Pumkin69 commented:
i know a lot of voxes who speak in third person so this will be great for them.

thetolbean commented:
Things have gone very wrong

PursuitInAshes reviewed (approved):
Code looks good.

I would really prefer this doesn't have a placeholder icon for the action, but I'm not sure what the icon could be.

1TALERD commented:
I don't see the point in not adding this. We definitely need it.

slarticodefast commented:
This will need to be adjusted to https://github.com/space-wizards/space-station-14/pull/35555 so it is copied correctly when the player is cloned.
Just add the component to the cloning settings in `clone.yml`

thetolbean commented:
> This will need to be adjusted to #35555 so it is copied correctly when the player is cloned. Just add the component to the cloning settings in `clone.yml`

I've gone ahead and updated `clone.yml` to include illeism
